### ISSUE TITLE
order of magnitude difference in rotation profiles at C depletion between 15140 and release candidate

### ISSUE THREAD

**Issue by mathren** [MESAHub/mesa#353]
I am finding a large change in the rotational profile (omega vs. mass) with two setups that should be comparable between 15140 and the release candidate. I am comparing models at core carbon depletion, starting with omega/omega_crit=0.5 at ZAMS, and adopting a "standard" Spruit-Tayler dynamo for angular momentum transport. The plots below show the case for a 30Msun starting with omega/omega_crit = 0.5 at ZAMS (see minimal working examples attached),  but qualitatively this behavior is the same also for 20 and 35Msun (same initial rotation).

With the release candidate I get this profile:
![image](https://user-images.githubusercontent.com/49067756/145279308-454da7b1-5325-4993-a387-83db6a1fea4c.png)

While the (unsatisfying) situation with 15140 was this (*note the different scale on the y-axis*):
![image](https://user-images.githubusercontent.com/49067756/145279653-facfe6d8-2c1a-4583-a1d7-eaaeecd5995d.png)

where different colors correspond to different combinations of `mesh_delta_coef` and `mesh_time_coef`. Note that these variations are larger than variations obtained changing the initial omega/omega_crit at constant resolution in 15140 (this is why I was originally running these tests). 

@orlox do you have any thoughts?

The main differences between the two setups are in the value of `varcontrol_target` (1d-4 in 15140, 2x lower in the release candidate), and the use of the `conv_vel` variable a la Arnett 1969 unsupported in the release candidate, which should really not matter here. Here is the output of a non-manual comparison between the minimal working examples attached below (and in case of differences for one option, the defaults in 15140 are also printed):

```
compare_all_workdir_inlists ./MWE_rc1 ./MWE_15140
...reading ./MWE_rc1/inlist_both star_job namelist
...reading ./MWE_rc1/inlist_extra star_job namelist
...reading ./MWE_15140/inlist_both star_job namelist
...reading ./MWE_15140/inlist_extra star_job namelist

&star_job
2: MWE_15140                  	change_conv_vel_flag=.true.                                       
1: MWE_rc1                    	missing
default                       	change_conv_vel_flag=.false.                                      

2: MWE_15140                  	change_initial_conv_vel_flag=.true.                                       
1: MWE_rc1                    	missing
default                       	change_initial_conv_vel_flag=.false.                                      

2: MWE_15140                  	new_conv_vel_flag=.true.                                       
1: MWE_rc1                    	missing
default                       	new_conv_vel_flag=.false.                                      

/ !end star_job namelist
...reading ./MWE_rc1/inlist_both eos namelist
...reading ./MWE_rc1/inlist_extra eos namelist
...reading ./MWE_15140/inlist_both eos namelist
...reading ./MWE_15140/inlist_extra eos namelist

&eos
/ !end eos namelist
...reading ./MWE_rc1/inlist_both kap namelist
...reading ./MWE_rc1/inlist_extra kap namelist
...reading ./MWE_15140/inlist_both kap namelist
...reading ./MWE_15140/inlist_extra kap namelist

&kap
/ !end kap namelist
...reading ./MWE_rc1/inlist_both controls namelist
...reading ./MWE_rc1/inlist_extra controls namelist
...reading ./MWE_15140/inlist_both controls namelist
...reading ./MWE_15140/inlist_extra controls namelist

&controls
1: MWE_rc1                    	varcontrol_target=0.0005                                       
2: MWE_15140                  	varcontrol_target=0.0001                                       

surf_omega_div_omega_crit_tol not in defaults
surf_omega_div_omega_crit_limit not in defaults
2: MWE_15140                  	surf_w_div_w_crit_limit=0.95                                         
1: MWE_rc1                    	missing
default                       	surf_w_div_w_crit_limit=0.99                                         

/ !end controls namelist
```

[MWE_rc1.zip](https://github.com/MESAHub/mesa/files/7678799/MWE_rc1.zip)
[MWE_15140.zip](https://github.com/MESAHub/mesa/files/7678869/MWE_15140.zip)

**Comment by mathren** [comment]
Here is also a plot of omega/omega_crit profiles at C depletion for two of the models above (using `mesh_delta_coeff = time_delta_coeff=0.8` in both models). The release candidate seems to reach omega/omega_crit >1 at the CO/He interface. The dots show the mesh spacing.

Maybe this can help figure out what is going on...

![image](https://user-images.githubusercontent.com/49067756/145312604-90c7b80e-7c1a-4c1e-806a-d8a8b8ac9ebe.png)

zoom in on the y-axis:
![image](https://user-images.githubusercontent.com/49067756/145312698-ee26637e-80e4-4e9d-a02b-a023e66fe3d8.png)

and omega(mass) on the same plot for these two models:
![image](https://user-images.githubusercontent.com/49067756/145312899-483236c0-1387-4cce-8728-23cdc9387b34.png)

**Comment by orlox** [comment]
Hi @mathren ,

I'm currently investigating this and have my own suspicions as for the source. However, would it be possible for you to provide a MWE that shows the same issue but runs a bit faster? For instance, do you need the very small varcontrol and to change the network from ZAMS for this problem to appear? Is the discrepancy apparent in an earlier evolutionary phase?

Also, a brief on-the-side recommendation from checking your inlists. It's generally better to modify initial conditions such as rotation and metallicity using relaxation routines rather than directly setting them. You can have large changes induced by a one-step change (your initial model goes significantly off thermal equilibrium) and this can produce your actual initial rotation to have important mismatches from what you initially chose. Probably entirely unrelated to your issue though.

**Comment by mathren** [comment]
I don't think the very high resolution is needed (though I don't think these models are resolved in terms of their rotational profile, which was my starting point with 15140). I would be very interested in knowing if you think that the issue with resolution and rotation rate are related to how I initialize the rotation rate. If it is such a bad idea to use the `set_initial_*`, maybe this should be made more explicit to the users (I'd suggest maybe to write this in the comments to `star_job.defaults` and/or `controls.defaults`). It seems many test cases use these options, so I probably am not the only user that copied from there. Thanks for the tip!

If it is ok to debug this with unresolved models you can definitely increase `mesh_delta_coeff` and `time_delta_coeff`, and possibly `varcontrol_target` too, I expect the problem to still show up. I have not tested this, but I suspect the nuclear reaction network here is not relevant, so using `basic.net` should suffice to reproduce this behavior.

In terms of an earlier stopping criterion, here is a plot from the models above of their total angular momentum as a function of (log) time (until C depl). The vertical dashed line marks the TAMS (X_center<0.001) and the dotted line core He depletion (Y_center <0.001). Clearly, the total angular momentum of the stars becomes significantly different during He core burning

![image](https://user-images.githubusercontent.com/49067756/145465140-bb4b0e8b-d692-4871-a6e2-b3b947e3d7f5.png)

Here is the same plot as a function of central temperature, which might be more useful to set a stopping criterion (something like max log_center_T = 8.1 might already be enough to see some difference)

![image](https://user-images.githubusercontent.com/49067756/145466236-9fca392f-1037-4a6b-9b1a-0db04f274b2d.png)

Unfortunately, before next week it will be very hard for me to develop a faster MWE
and verify that the issue still occurs, I hope these plots can help in the meantime.

**Comment by mathren** [comment]
Also, I had a grid running already with the setup from MWE_rc1.zip varying only initial masses, `mesh_delta_coeff` (labeled by m in this plot) and `time_delta_coeff` (labeled t), so here is the partial result of this resolution test.

From the y-axis values, I see that this issue seems to happen across the mass range 10-40Msun at least. From the differences in the curves, I think the "non-resolved-ness" of the rotation profile has gotten worse than in 15140 (where I was already not finding converged results for M>10Msun). The resolution issue is of course secondary at this point, but I thought it might be worth bringing it up.

These models all start with omega_div_omega_crit=0.5 not "relaxed" but "set" at ZAMS, and the profiles are plotted at C depletion:
![image](https://user-images.githubusercontent.com/49067756/145481018-28c08603-41e1-4134-b655-48d90970513e.png)

**Comment by orlox** [comment]
This is very likely a consequence of the removal of time smoothing for D_ST. I re-ran the MWE for 15140, but turning off this time smoothing. Attached is what I get, an even larger difference in final properties just by changing this in 15140. I lowered time and spatial resolution significantly to compute this rapidly, so that might be the reason for the even larger difference to what you initially reported @mathren . I'll work on restoring time smoothing and let you know.
[angular_momentum_evolution.pdf](https://github.com/MESAHub/mesa/files/7697817/angular_momentum_evolution.pdf)
[inlist_both.zip](https://github.com/MESAHub/mesa/files/7697820/inlist_both.zip)

**Comment by orlox** [comment]
FYI, axes labels are flipped in the plot above!

**Comment by orlox** [comment]
The 30 Msun, 0.5 critical case in particular is a bit hard to work with as it gives partially chemically homogeneous evolution which is indeed challenging to get to a convergent state. Instead I did a few tests with the 30 Msun star but starting at 0.1 critical. After reimplementing time-smoothing I did a few tests, both with the new version and 15140, and with and without time-smoothing. I find that with a given choice of smoothing the results match nicely across versions. So I've set the defaults to use the same parameters for smoothing from the previous version. Made a PR and this should be in the next release.

Thanks for catching this @mathren !

For reference I attach the work folders I used, these use the latest version of the pm/rotation_fixes branch.
[final_tests.tar.gz](https://github.com/MESAHub/mesa/files/7706236/final_tests.tar.gz)
![comparison](https://user-images.githubusercontent.com/537248/145872546-8db25bbe-e933-4b0e-b93a-00537d62a5dd.png)
